### ISSUE TITLE
Mobile Push Pt4: magic deeplink

### ISSUE THREAD

**Issue by davismtl** [mozilla/fxa-features#52]
When deeplinking into the app via an Adjust link sent via SMS, auto-login the user instead of only pre-filing the email field.

This is phase 2 of: 
https://waffle.io/mozilla/fxa-features/cards/5820a9739ca711d4005a5408

[Add Feature Doc Link HERE]

**Comment by seanmonstar** [comment]
The experience would be super slick! Some obstacles that come to mind are:
- [SMS in insecure](https://www.wired.com/2016/06/hey-stop-using-texts-two-factor-authentication/). It's possible to intercept SMS messages, which makes it possible for someone to intercept it and automatically be logged in to someone else's account.
- It's also scary to auto login someone to an account, if they clicked on the wrong link somehow, a user would have setup Firefox, and may not realize they are now syncing their history and passwords into someone else's account.

Perhaps the UI could help prevent the second, by pointing out in big letters what account you've just signed in as. An alternative @rfk mentioned was maybe they are prompted to enter a short code that is displayed on their desktop. Much easier and faster than typing email/password, not quite as slick as "woah, I'm already signed in", but perhaps safer.

**Comment by davismtl** [comment]
Excellent points @seanmonstar ! I'm happy to say that @shane-tomlinson, @vladikoff and @vbudhram participated in the brainstorm of this feature while we were in Toronto so we had some time to think about it a bit. I think we can ultimately find a way to secure the process and keep it slick.

The auto-login would work by passing the short code proposed by @rfk (and @vladikoff @vbudhram ) via the deeplink. We'd be removing the friction of entering it manually and having to go back and forth between devices.

Solutions we thought of to mitigate the lack of security of SMS:
- Short lived links
- One time use
- Display message on originating desktop that says that device was successfully added (if they didn't add yet, send them to device manager)
- Or, display an authorization message on originating device when new device is requesting to be added via "magic link".
- If IP of second device is in different location, perhaps we still use an email confirmation (although ideally we would want to avoid this since we add more friction back into the flow)

Other benefits we thought of for the "magic link":
- UA aware and can customize flow accordingly.
- Can detect if app is installed yet and send you through install flow first before login (Adjust and Branch can do that, normally)
- If already installed, just does login
- Can detect if an account is already signed in

**Comment by rfk** [comment]
> It's also scary to auto login someone to an account

Indeed!  To be secure against fixation attacks, this would need to show a confirmation screen before doing the login.  I think the right way to think about it is, instead of having the login page show a username+password field, it shows a message "Sign in to xyz@example.com?" and a big blue "Sign in" button for the user to click.  So it's "automatic" in the sense that you don't have to type anything, you just have to click the button.

> maybe they are prompted to enter a short code that is displayed on their desktop

I also like this option as an additional layer of security, that maybe we can deploy as an intermediate step.

**Comment by shane-tomlinson** [comment]
> "Sign in to xyz@example.com?" and a big blue "Sign in" button for the user to click. So it's "automatic" in the sense that you don't have to type anything, you just have to click the button.

Yes, this was part of the idea that I presented. There needs to be two components to securely pairing the devices, 1) security for the new devices, 2) security for devices already connected.

This can be handled in a couple of ways.
1. A JPAKE code copied from one device to another. This operates on the assumption that only someone with access to both devices can pair.
2. Confirmation screens on both devices where the user on both devices must take an action for the devices to be paired, probably a confirmation button. The user on the new device would see `Hey, you are connecting as xyz@foo.com`, and the user on the existing device would see `Do you want to allow user on <device type, e.g., Firefox for Android> in <location, e.g., London> to connect?`
3. Some other awesome thing we haven't considered.

My personal preference is to do away with typing if at all possible. We did JPAKE with old sync and from what I understand users found it somewhat confusing. Maybe the UX problem is the context in which it was presented, maybe we can do a better job. But, I have this question if we can do away with typing anything using DH key exchange, with as dangerous as that sounds in the browser.

Basic flow:
1. User decides to pair a device. On device A, user enters phone number of device, sends SMS. SMS contains link to install Firefox and a "pairing code".
2. Device B receives SMS, opens pairing page. User is asked if they want to pair with the other device. On Device A, user is asked if they want to allow Device B to connect. If confirmations occur on both devices, the "pairing code" is used to facilitate DH key exchange between the two devices.
3. User on device A generates creds for device B. Creds are encrypted using the keys exchanged between devices. Creds are sent to device B.
4. Device b decrypts creds, uses them as its own, starts syncing.

**Comment by vbudhram** [comment]
> My personal preference is to do away with typing if at all possible.

I agree. I personally like the idea of the one time use login link with an expiration. A sample flow could be
1. On account verification, generate a login code that expires in X time. Send as auth-server response when verification occurs.
2. On email confirmed screen (Desktop), display success and option to SMS invite link or email link.
3. On email confirmed screen (Mobile), display success and option to `Continue` using same link sent via SMS/email.
4. On Mobile, upon opening link, user is taken to corresponding App Store using Adjust, after install, Adjust relaunches app.
5. On app launch, user is taken directly to a confirm screen `Hey, you are connecting as xyz@foo.com, continue?`, user clicks continue and is signed in.
6. App introduction/guide screen is shown.

This could greatly help reduce the friction of having to approve the login on the Desktop or enter in a paring code. We could also have some additional checks to see if user is logging in from known location/IP/OS etc, that would invalidate code and have them sign-in manually.

**Comment by shane-tomlinson** [comment]
> On app launch, user is taken directly to a confirm screen Hey, you are connecting as xyz@foo.com, continue?, user clicks continue and is signed in.

My personal belief is after this step, device A should also explicitly authorize pairing with device B. Neither email nor SMS are all that secure, both can be intercepted, and I wouldn't want some jerk with a Stingray pairing as me and stealing my passwords.

**Comment by rfk** [comment]
> We did JPAKE with old sync and from what I understand users found it somewhat confusing.
> Maybe the UX problem is the context in which it was presented, maybe we can do a better job. 

It's also possible that users are more comfortable with this sort of flow these days, especially in a desktop<->mobile setting.  If they've ever set up a Roku or a Chromecast or similar, they've done a pairing flow before.

The sync pairing code was also incredibly long, we should be able to use additional info in the link itself to allow a shorter pairing code.

**Comment by seanmonstar** [comment]
> It's also possible that users are more comfortable with this sort of flow these days, especially in a desktop<->mobile setting. If they've ever set up a Roku or a Chromecast or similar, they've done a pairing flow before.

It also seems to me that the user is likely right in front of the desktop when they are doing this. The user just finished signing in to Firefox desktop, and send "sure, send me a link to my phone". They get the SMS, and take care of it right there. Prompting them "hey, look at Firefox on your desktop for a code (or confirmation button or whatever)" seems like it should be easy in their context.

There is a possibility that the sending of the SMS takes a long time, and they then try to act on it when they are away from their desktop, I suppose.

**Comment by shane-tomlinson** [comment]
> Prompting them "hey, look at Firefox on your desktop for a code (or confirmation button or whatever)" seems like it should be easy in their context.

This is what I hoped too, which is why it seems acceptable to me to have some explicit authorization model, either via a code or button, or some other magic like a QR code scan.

**Comment by davismtl** [comment]
FEATURE CARD MOVED TO NEW PRODUCT BOARD. ADD NEW COMMENTS HERE: 
https://projects.growthhackers.com/s2/xZy2i6N58NewRjGrKdINYA
